portal/models/reporting.py

```diff
@@ -102,6 +106,120 @@ def get_reporting_stats():
     return stats
 
 
+def adherence_report(
+        requested_as_of_date, acting_user_id, include_test_role, org_id,
+        response_format, celery_task):
+    """Generates the adherence report
+
+    Designed to be executed in a background task - all inputs and outputs are
+    easily serialized (executing celery_task parent an obvious exception).
+
+    :param requested_as_of_date: string form of as_of_date, or None to use now
+    :param acting_user_id: id of user evoking request, for permission check
+    :param include_test_role: set to include test patients in results
+    :param org_id: set to limit to patients belonging to a branch of org tree
+    :param response_format: 'json' or 'csv'
+    :param celery_task: used to update status when run as a celery task
+    :return: dictionary of results, easily stored as a task output, including
+       any details needed to assist the view method
+
+    """
+    acting_user = User.query.get(acting_user_id)
+    if requested_as_of_date:
+        as_of_date = FHIR_datetime.parse(requested_as_of_date)
+    else:
+        as_of_date = datetime.utcnow()
+
+    # If limited by org - grab org and all it's children as filter list
+    requested_orgs = (
+        OrgTree().here_and_below_id(organization_id=org_id) if org_id
+        else None)
+
+    patients = patients_query(
+        acting_user=acting_user,
+        include_test_role=include_test_role,
+        requested_orgs=requested_orgs)
+    data = []
+    current, total = 0, patients.count()
+    for patient in patients:
+
+        # occasionally update the celery task status if defined
+        current += 1
+        if not current % 25 and celery_task:
+            celery_task.update_state(
+                state='PROGRESS', meta={'current': current, 'total': total})
+
+        if len(patient.organizations) == 0:
+            # Very unlikely we want to include patients w/o at least
+            # one org, skip this patient
+            continue
+
+        try:
+            acting_user.check_role('view', other_id=patient.id)
+        except Unauthorized:
+            # simply exclude any patients the user can't view
+            continue
+
+        qb_stats = QB_Status(user=patient, as_of_date=as_of_date)
+        row = {
```

Comment:
Would it make sense to start to formalize some of the CSVs we're writing as classes?

Reply:
err - perhaps.  i think i'll punt on that level of change at this time, especially as this needs serialization support during its redis life.